server/src/main/java/org/elasticsearch/index/engine/LuceneSyntheticSourceChangesSnapshot.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.FieldDoc;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.util.ArrayUtil;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.IndexVersion;
+import org.elasticsearch.index.fieldvisitor.LeafStoredFieldLoader;
+import org.elasticsearch.index.fieldvisitor.StoredFieldLoader;
+import org.elasticsearch.index.mapper.MappingLookup;
+import org.elasticsearch.index.mapper.SourceFieldMetrics;
+import org.elasticsearch.index.mapper.SourceLoader;
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link SearchBasedChangesSnapshot} that utilizes a synthetic field loader to rebuild the recovery source.
+ * This snapshot is activated when {@link IndexSettings#RECOVERY_USE_SYNTHETIC_SOURCE_SETTING}
+ * is enabled on the underlying index.
+ *
+ * The {@code maxMemorySizeInBytes} parameter limits the total size of uncompressed _sources
+ * loaded into memory during batch retrieval.
+ */
+public class LuceneSyntheticSourceChangesSnapshot extends SearchBasedChangesSnapshot {
+    public static final long DEFAULT_MEMORY_SIZE = 4 * 1024 * 1024; // 4MB
```

Comment:
> If I understand correctly, we expect that rebuilding the source using synthetic source produces a byte[] that is larger than the original compressed source, and that's why we want to limit the memory usage with a limit.

I think this more about how many operations can be in memory at a time before loading more operations (as part of `LuceneSyntheticSourceChangesSnapshot#nextOperation(...)`. The idea is that loading more operations at a time reduces the overall cost of synthesizing _source, but we should be careful with memory usage. So we load a batch until `searchBatchSize` or `maxMemorySizeInBytes` limits are met. Then return to caller.

In case of CCR I think it works well with its own `maxBatchSize` in `ShardChangesAction#getOpertations(...)`? By default we read upto 4MB worth of operations and then buffer that locally in that method and then read the next batch of operations. It is like a limit on a different level? I think in the worst situation we may endup using CCR max batch size + LuceneSyntheticSourceChangesSnapshot.maxMemorySizeInBytes of memory?

Reply:
> It is like a limit on a different level? 

I think they share some similarity in that they can be both used to control the total size of operations to load in memory on the leader cluster. By batching up to 4MiB at the changes snapshot level, we silently increase the CCR max. batch possible size of 12.5% while preventing `max_read_request_size` lower than 4MiB to have an effect on the leader cluster. 

It's not a blocker though, but I think we should fix that. One solution could be to pass down some predicate from the `newChangesSnapshot()` method down to the `LuceneSyntheticSourceChangesSnapshot` instance that uses it to know if it can load the next doc source.